cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -309,6 +348,21 @@ std::shared_ptr<ScalarFunction> MakeArithmeticFunctionNotNull(std::string name,
   return func;
 }
 
+template <typename Op>
+std::shared_ptr<ScalarFunction> MakeScalarArithmeticFunction(std::string name,
+                                                             const FunctionDoc* doc) {
+  auto func = std::make_shared<ScalarFunction>(name, Arity::Unary(), doc);
+  // 8-bit signed integer
+  // ArrayKernelExec exec = ScalarUnary<Int8Type, Int8Type, Op>::Exec;
+  // DCHECK_OK(func->AddKernel({int8()}, int8(), exec));
+
+  for (const auto& ty : NumericTypes()) {
+    auto exec = NumericEqualTypesBinary<ScalarUnaryType, Op>(ty);
```

Comment:
It's fortunate that NumericEqualTypesBinary can be reused here, but we should rename it to avoid confusion. Maybe `ArithmeticExecFromOp`?
```suggestion
    auto exec = ArithmeticExecFromOp<ScalarUnary, Op>(ty);
```

Reply:
Also, note that although `ScalarUnary` is acceptable for "negate", "negate_checked" will need `ScalarUnaryNotNull` to ensure that we don't generate spurious errors when looking at values "under" null bits